### DISCUSSION TITLE
List of ideas

### DISCUSSION THREAD

**Discussion by Python-simulation** [qcha41/usit#12] [Ideas]
Example:
- [ ] Idea to think about
- [x] Idea to do (no necessary done)
- [ ] ~~Idea rejected~~

List:
- [x] allow to change measure step to set step to allow workflow: copy+paste+change
- [x] Add internal functions by right clicking on recipe. Need wait step and pause step (can already pause with param_unit='user-input' in an action of a driver)
- [x] Allow to hide curve in scanner (hide scan2 but see scan1 and scan3 -> need variable attached to Dataset for toggling) (could also remove data like plotter)
- [ ] Add color option to plot in scanner (seems a priori complicated to manage easily in GUI) Look at pyqtgraph example: 'Line Plot update' (could also attached variable to Dataset for color (alpha would still be conditioned from scan id))
- [ ] Tell user if want to recover data if close scanner without saving or crash (just last data?)
- [ ] Stop using local tmp folder to use custom one. With auto cleanup?
- [ ] See drivers from https://github.com/AlexShkarin/pyLabLib
- [ ] Look at https://micro-manager.org/ for ideas/drivers
- [ ] ~~Allow multiple scanner~~ (too risky for drivers in my opinion and too complicated at the time. Would required complete restructure of scan)
- [ ]  ~~Add option for data format saving (sep= \t space, comma...header format...)~~

**Comment by FateDecider** [comment]
A few suggestions:

1. May I suggest to set a default configuration to import in - when scanner is opened? There are times the various devices might glitch and one has to restart the GUI, and reimport the desired configuration. Other times, I would more heavily use one configuration file for a few days.

2. For each configuration file, I would like to have a default measured variable to plot on the right. This is very useful for N-dimensional scans, where a lot of things are measured, but some of these measurements are just for sanity checks and not crucial to the main experiment.

3. Would be nice to have the 'pause' step implemented in system driver. Instead of specifying a wait time, I am thinking of pausing the scanner during the automatized (N-dimensional scan) recipe. This is particularly useful for certain variables that cannot be controlled by the computer - like manual translational stages. It is hard to estimate how much time to wait, when manually changing the variable.

4. When plotting measured N-dimensional arrays in scanner, it will be nice if user can specify plot labels. Maybe this is best done when plotting dataframes? Also, would be nice to re-update the x-axis label of the fourier-transformed data.

**Reply by Python-simulation** [reply]
1. I have thought about this. But currently, importing a configuration starts the instruments used in it and I don't think this should happen by default when opening a scanner (perhaps a list of recently used configurations would be better, as in a lot of software).

2. A few commits ago, I changed the way variables in a scan are plotted, so that now the last plotted variable remains plotted even when multiple recipes are used. Previously, it would jump from one recipe to another to display the active one. This may be enough for you if you repeat the same scan several times and don't want to change the plotted variable. But if you really want a specific plotted variable to be remembered, in that case I don't think it's useful enough to implement. But if someone thinks it's useful and wants to implement it, do you want a default x and y for each recipe in a scan (recipe_1: x=var1, y=var2; recipe_2: x=var5, y=var11)?

3. I see, I've used a delay in some experiments and it's true that it's not the best way to wait for an user input in a scan. I'll have to think about it because the way the scan is implemented makes it complicated to stop it from the inside (I think).

4. As arrays have no labels, the labels used in the plot are only temporary to select data and are not saved in the file. Plotting dataframes uses existing labels and should not be modified, as they are generated by the driver. You can modify them once saved or in the driver. But this is done on the user's side. What do you mean by "update x axis"? I don't see any problem with the FFT and the x-axis. Don't forget that autolab is mainly used for doing experiments, not for analyzing data.

**Reply by FateDecider** [reply]
1. I see, that is true, I would agree on this. Your suggestion could nevertheless be a nice feature.

2. Yes, what I had in mind is what you have described. Recipe_1: x = var1, y = var2; recipe_2: x = var5, y = var11

3. Yes I agree it would be hard to stop the Scanner from the inside. An easy, but less user-friendly, workaround would be to define a recipe step that requires user to key in (y/n) on the console, akin to pausing the scan.

4. I agree autolab is mainly for doing experiments, but would be nice, if FFT alone could have the x-axis label updated to frequency. Currently, what I have is this (where FFT and log y has been checked):
![image](https://github.com/autolab-project/autolab/assets/57542293/458e7c06-c3b6-4726-aa57-b4ae2263e75a)

**Reply by Python-simulation** [reply]
1. Will add in the future a menu for the last used configs.
2. I will keep it in mind but it is honestly not on my top list priority.
3. I will see if using `input` is already possible in a driver or if it needs to be implemented using a specific variable type.
4. I don't think this is possible because the initial label depends on the user. If I have position (mm) and do FFT, what should be the new label? Only `FFT(initial_label)` or keep `initial_label` are straight forward enough. And should be the new unit? Can't deduced it easily.

**Reply by Python-simulation** [reply]
3. I tested input and it works. But this is not a supported way of doing it as it involved interacting with the console (and uses input that is buggy in Spyder). I will see if I can add a pause internally to the scanner.

But for now you can do this.
In any driver, define:
```
    def get_pause(self):
        return str(input('Press enter to continue'))
```
and in get_driver_model:
```
        model.append({'element': 'variable', 'name': 'pause', 'type': str,
                      'read': self.get_pause,
                      'help': 'Wait for an user input (Press enter to continue)'})
```
Then just press enter on the console to release the input function.

**Reply by Python-simulation** [reply]
I added a new feature in branch v2.0rc1to ask user for input during a scan. The dummy driver has been updated to test this feature.
In a driver you can define an action as follow:
```
    def user_input(self, value):
        print('User input:', value)
```
```
        model.append({'element': 'action', 'name': 'user_input',
                      "param_type": str, "param_unit": "user-input",
                      'do': self.user_input,
                      'help': 'Asks the user for input (can be used to pause scan until user presses enter)'})
```

**Comment by Sky-creates** [comment]
Suggestions related to monitoring
When monitoring a numpy 2D array as an image, can we
- Adjust the x range and y range during monitoring? Currently, after zooming into a certain range, the x and y range will automatically reset after the new data is taken.
- Have a tool to draw basic shapes such as lines/circles on the image?
Thanks!

**Reply by Sky-creates** [reply]
- Also, adjusting the value range in addition to x/y range would be helpful

**Reply by Python-simulation** [reply]
I made the necessary changes (not committed yet) to zoom on an image without refreshing the axes range on new data acquisition.
The axes can also be adjusted independently thanks to a checkbox to disable the fixed ratio of 1.

I'm working on adding circles, rectangles and lines. For now I can add and move them but not yet resize them.

**Reply by Python-simulation** [reply]
Done in fefe2f7 and published in [pypi](https://pypi.org/project/autolab/2.0.1/)

**Comment by Sky-creates** [comment]
I am thinking is it possible to allow using save button inside the scanner recipe? I can think of the following scenarios that this might be useful
1. use of customized file names
2. For 2D scans, saving the results everytime a 1D scan is completed helps to prevent data loss if anything crashes in the middle of measurement.

**Reply by Python-simulation** [reply]
In the curent state, I don't see a simple solution to achieve this.
- If needed, you can save your data at anytime during a scan by clicking on the save button.
- No data is ever lost in case of a crash as the scan data is saved at every step in a temporary folder. By default it is located in %tmp% but you can change the default folder in the autolab settings.

**Reply by Sky-creates** [reply]
I see. Thank you! Maybe I can think about adding an action to periodically copy files out from the temp folder.

**Reply by FateDecider** [reply]
@Python-simulation I can attest %temp% is not a good solution... because my computer auto-updated (i have already postponed it for the longest duration possible...) during an overnight measurement

**Reply by Python-simulation** [reply]
I had similar problems myself. This is why I added the option to change the default temp folder to a custom one. I think it would be safer in futur version to add and set a custom temp folder in the autolab local folder instead of %temp%. In the mean time you can change it in the preferences panel if you use autolab 2.x

**Comment by Sky-creates** [comment]
When plotting a 2D image, what is the best way to get the XY coordinate of the point where my mouse is pointing at? When adding an geometry (such as a circle), is it able to read out the center position of the circle? WIth these functionalities I will be able to replace the circle to the exact same location everytime I restart the monitor window. Thanks!

**Reply by Python-simulation** [reply]
Could be a nice feature indeed. There is currently no way to get the coordinate. I personally don't have time to implement this but if someone wants to do it, they will need to look into GUI_utilities.py mouseMoved and see how to use it for the image (myImageView). Could also look into add_shape to add a coordinate option and create methods to change the figure according to the entered coordinates.
